internal/component/faro/receiver/rate_limiter.go

```diff
@@ -0,0 +1,174 @@
+		}),
+		rateLimitDecisions: promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "faro_receiver_rate_limiter_requests_total",
+				Help: "Total number of requests processed by the rate limiter per app/environment.",
+			},
+			[]string{"app", "env", "allowed"},
+		),
+	}
+}
+
+// GetPoolLimiter returns the rate limiter for the given key.
+// Returns the limiter and a boolean indicating if it exists.
+func (r *AppRateLimitingConfig) GetPoolLimiter(key AppRateLimitingConfigKey) (*rate.Limiter, bool) {
+	appLimiter, exists := r.pool[key]
+	if !exists {
+		return nil, false
+	}
+	return appLimiter.limiter, true
+}
+
+// SetPoolLimiter sets a rate limiter for the given key and returns it.
+func (r *AppRateLimitingConfig) SetPoolLimiter(key AppRateLimitingConfigKey, limiter *rate.Limiter) *rate.Limiter {
+	r.pool[key] = &AppRateLimiter{
+		limiter:  limiter,
+		lastUsed: time.Now(),
+	}
+	r.activeApp.Set(float64(len(r.pool)))
+
+	return limiter
+}
+
+// CleanupRoutine starts a goroutine that periodically removes inactive rate limiters.
+// It prevents memory leaks by deleting limiters that haven't been used within the expiry duration.
+// The routine runs until the context is cancelled.
+func (r *AppRateLimitingConfig) CleanupRoutine(ctx context.Context, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			r.cleanupExpiredLimiters()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// cleanupExpiredLimiters removes rate limiters that haven't been used recently.
+func (r *AppRateLimitingConfig) cleanupExpiredLimiters() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	now := time.Now()
+	removed := 0
+	for key, appLimiter := range r.pool {
+		if now.Sub(appLimiter.lastUsed) > DEFAULT_LIMITER_EXPIRY {
+			delete(r.pool, key)
+			removed++
+		}
+	}
+
+	// Update pool size metric if limiters were removed
+	if removed > 0 {
+		r.activeApp.Set(float64(len(r.pool)))
+	}
+}
+
+// Allow checks if a request is allowed for the given app/environment combination.
+// Creates a new rate limiter if one doesn't exist for the app/env key.
+// Updates the lastUsed timestamp on each call to prevent the cleanup routine
+// from removing active limiters.
+func (r *AppRateLimitingConfig) Allow(app, env string) bool {
+	key := ParseAppRateLimitingConfigKey(app, env)
+
+	r.mu.Lock()
+	limiter, exists := r.GetPoolLimiter(key)
+	if !exists {
+		newLimiter := r.getNewLimiter()
+
+		limiter = r.SetPoolLimiter(key, newLimiter)
+	} else {
+		// Limiter exists, update its last used timestamp to prevent cleanup
+		appLimiter := r.pool[key]
+		appLimiter.lastUsed = time.Now()
+	}
+
+	r.mu.Unlock()
+
+	allowed := limiter.Allow()
+	r.rateLimitDecisions.WithLabelValues(app, env, fmt.Sprintf("%t", allowed)).Inc()
+	return allowed
+}
+
+func (r *AppRateLimitingConfig) getNewLimiter() *rate.Limiter {
+	t := time.Now().Add(-time.Duration(float64(time.Second) * float64(r.rate) * float64(r.burst)))
```

Comment:
@hhertout could you please explain a bit why time is calculated like this? why not just time.Now()? Maybe adding a comment above would be great